### ISSUE TITLE
Project Configuration Question

### ISSUE THREAD

**Issue by derpeter** [spdx/spdx-gradle-plugin#49]
I'm struggeling getting the plugin running correctly as im not that much of an gradle expert.

I added the provided example configuration to the project gradle and the json file is generated as expected. 

But it only contains the information on the project it self not the depencies of the application. So i guess i misst something important here. 

Gradle did not allow me to set configuration set  with the message 

`Gradle DSL method not found: 'listOf()'" 
`
Can this be the cause ?

I also tried to move the plugin configuration to the app level gradle file but gradle complains

`Cannot locate tasks that match ':spdxSbomForRelease' as task 'spdxSbomForRelease' not found in root project 'myAPP'.
`

Sorry if this is stupid question but maybe it helps to improve documentation

**Comment by derpeter** [comment]
Project Level Gradle


buildscript {
    repositories {
        google()
        mavenCentral()
    }
    dependencies {
        classpath 'com.android.tools.build:gradle:8.1.2'
        def nav_version = "2.5.3"
        classpath "androidx.navigation:navigation-safe-args-gradle-plugin:$nav_version"
        classpath 'org.jetbrains.kotlin:kotlin-gradle-plugin:1.6.21'
    }
}



plugins {
    id("org.spdx.sbom") version "0.2.0"
}

configurations {
    runtimeClasspath
}



task clean(type: Delete) {
    delete rootProject.buildDir
}

**Comment by loosebazooka** [comment]
Oh, if this is an android project, it could be that android uses some non-default configurations for defining dependencies?

**Comment by derpeter** [comment]
It is an Android Projekt. (currently in the process of beeing opensourced, that why i need the depencies ) 
Running  ./gradlew dependencie also results in an empty list but  ./gradlew app:dependencie seems to work.

What do you mean by available configurations ?

**Comment by loosebazooka** [comment]
Fyi codeblocks in github markdown are triple backticks `` ``` ``

Yeah so if you want to set the configuration to "debugRuntimeClasspath" use the config
```
      configurations.set(listOf("debugRuntimeClasspath"))
```

I don't actually know enough about how gradle is used in android but I doubt that the debug variant is the one you generally publish. Isn't there a more publishable variant? Can you paste the whole output of the following command which lists out all the configurations

```
./gradlew --console plain dependencies | fgrep ' - '
```

**Comment by derpeter** [comment]
> Fyi codeblocks in github markdown are triple backticks `` ``` ``
> 

A sorry i used the button above the editor.

> Yeah so if you want to set the configuration to "debugRuntimeClasspath" use the config
> 
> ```
>       configurations.set(listOf("debugRuntimeClasspath"))
> ```
> 

It seems this only working with kotlin android project, for java it looks like

configurations = ['debug']


> I don't actually know enough about how gradle is used in android but I doubt that the debug variant is the one you generally >publish. 

Your right, i also tried release with the same results

> Isn't there a more publishable variant? Can you paste the whole output of the following command which lists out all the >configurations
> 
> ```
> ./gradlew --console plain :app:dependencies | fgrep ' - '
> ```

```
./gradlew --console plain :app:dependencies | fgrep ' - '
androidApis - Configuration providing various types of Android JAR file
androidJdkImage - Configuration providing JDK image for compiling Java 9+ sources
androidTestAnnotationProcessor - Classpath for the annotation processor for 'androidTest'. (n)
androidTestCompileOnly - Compile only dependencies for 'androidTest' sources. (n)
androidTestDebugAnnotationProcessor - Classpath for the annotation processor for 'androidTestDebug'. (n)
androidTestDebugCompileOnly - Compile only dependencies for 'androidTestDebug' sources. (n)
androidTestDebugImplementation - Implementation only dependencies for 'androidTestDebug' sources. (n)
androidTestDebugRuntimeOnly - Runtime only dependencies for 'androidTestDebug' sources. (n)
androidTestDebugWearApp - Link to a wear app to embed for object 'androidTestDebug'. (n)
androidTestImplementation - Implementation only dependencies for 'androidTest' sources. (n)
androidTestReleaseAnnotationProcessor - Classpath for the annotation processor for 'androidTestRelease'. (n)
androidTestReleaseCompileOnly - Compile only dependencies for 'androidTestRelease' sources. (n)
androidTestReleaseImplementation - Implementation only dependencies for 'androidTestRelease' sources. (n)
androidTestReleaseRuntimeOnly - Runtime only dependencies for 'androidTestRelease' sources. (n)
androidTestReleaseWearApp - Link to a wear app to embed for object 'androidTestRelease'. (n)
androidTestRuntimeOnly - Runtime only dependencies for 'androidTest' sources. (n)
androidTestUtil - Additional APKs used during instrumentation testing.
androidTestWearApp - Link to a wear app to embed for object 'androidTest'. (n)
annotationProcessor - Classpath for the annotation processor for 'main'. (n)
api - API dependencies for 'main' sources. (n)
compileOnly - Compile only dependencies for 'main' sources. (n)
coreLibraryDesugaring - Configuration to desugar libraries
debugAndroidTestAnnotationProcessorClasspath - Resolved configuration for annotation-processor for variant: debugAndroidTest
debugAndroidTestCompileClasspath - Resolved configuration for compilation for variant: debugAndroidTest
debugAndroidTestRuntimeClasspath - Resolved configuration for runtime for variant: debugAndroidTest
debugAnnotationProcessor - Classpath for the annotation processor for 'debug'. (n)
debugAnnotationProcessorClasspath - Resolved configuration for annotation-processor for variant: debug
debugApi - API dependencies for 'debug' sources. (n)
debugApiElements - API elements for debug (n)
debugCompileClasspath - Resolved configuration for compilation for variant: debug
debugCompileOnly - Compile only dependencies for 'debug' sources. (n)
debugImplementation - Implementation only dependencies for 'debug' sources. (n)
debugReverseMetadataValues - Metadata Values dependencies for the base Split
debugRuntimeClasspath - Resolved configuration for runtime for variant: debug
debugRuntimeElements - Runtime elements for debug (n)
debugRuntimeOnly - Runtime only dependencies for 'debug' sources. (n)
debugUnitTestAnnotationProcessorClasspath - Resolved configuration for annotation-processor for variant: debugUnitTest
debugUnitTestCompileClasspath - Resolved configuration for compilation for variant: debugUnitTest
debugUnitTestRuntimeClasspath - Resolved configuration for runtime for variant: debugUnitTest
debugWearApp - Link to a wear app to embed for object 'debug'. (n)
debugWearBundling - Resolved Configuration for wear app bundling for variant: debug
implementation - Implementation only dependencies for 'main' sources. (n)
lintChecks - Configuration to apply external lint check jar
lintPublish - Configuration to publish external lint check jar
releaseAnnotationProcessor - Classpath for the annotation processor for 'release'. (n)
releaseAnnotationProcessorClasspath - Resolved configuration for annotation-processor for variant: release
releaseApi - API dependencies for 'release' sources. (n)
releaseApiElements - API elements for release (n)
releaseCompileClasspath - Resolved configuration for compilation for variant: release
releaseCompileOnly - Compile only dependencies for 'release' sources. (n)
releaseImplementation - Implementation only dependencies for 'release' sources. (n)
releaseReverseMetadataValues - Metadata Values dependencies for the base Split
releaseRuntimeClasspath - Resolved configuration for runtime for variant: release
releaseRuntimeElements - Runtime elements for release (n)
releaseRuntimeOnly - Runtime only dependencies for 'release' sources. (n)
releaseUnitTestAnnotationProcessorClasspath - Resolved configuration for annotation-processor for variant: releaseUnitTest
releaseUnitTestCompileClasspath - Resolved configuration for compilation for variant: releaseUnitTest
releaseUnitTestRuntimeClasspath - Resolved configuration for runtime for variant: releaseUnitTest
releaseWearApp - Link to a wear app to embed for object 'release'. (n)
releaseWearBundling - Resolved Configuration for wear app bundling for variant: release
runtimeOnly - Runtime only dependencies for 'main' sources. (n)
testAnnotationProcessor - Classpath for the annotation processor for 'test'. (n)
testCompileOnly - Compile only dependencies for 'test' sources. (n)
testDebugAnnotationProcessor - Classpath for the annotation processor for 'testDebug'. (n)
testDebugCompileOnly - Compile only dependencies for 'testDebug' sources. (n)
testDebugImplementation - Implementation only dependencies for 'testDebug' sources. (n)
testDebugRuntimeOnly - Runtime only dependencies for 'testDebug' sources. (n)
testDebugWearApp - Link to a wear app to embed for object 'testDebug'. (n)
testFixturesAnnotationProcessor - Classpath for the annotation processor for 'testFixtures'. (n)
testFixturesApi - API dependencies for 'testFixtures' sources. (n)
testFixturesCompileOnly - Compile only dependencies for 'testFixtures' sources. (n)
testFixturesDebugAnnotationProcessor - Classpath for the annotation processor for 'testFixturesDebug'. (n)
testFixturesDebugApi - API dependencies for 'testFixturesDebug' sources. (n)
testFixturesDebugCompileOnly - Compile only dependencies for 'testFixturesDebug' sources. (n)
testFixturesDebugImplementation - Implementation only dependencies for 'testFixturesDebug' sources. (n)
testFixturesDebugRuntimeOnly - Runtime only dependencies for 'testFixturesDebug' sources. (n)
testFixturesDebugWearApp - Link to a wear app to embed for object 'testFixturesDebug'. (n)
testFixturesImplementation - Implementation only dependencies for 'testFixtures' sources. (n)
testFixturesReleaseAnnotationProcessor - Classpath for the annotation processor for 'testFixturesRelease'. (n)
testFixturesReleaseApi - API dependencies for 'testFixturesRelease' sources. (n)
testFixturesReleaseCompileOnly - Compile only dependencies for 'testFixturesRelease' sources. (n)
testFixturesReleaseImplementation - Implementation only dependencies for 'testFixturesRelease' sources. (n)
testFixturesReleaseRuntimeOnly - Runtime only dependencies for 'testFixturesRelease' sources. (n)
testFixturesReleaseWearApp - Link to a wear app to embed for object 'testFixturesRelease'. (n)
testFixturesRuntimeOnly - Runtime only dependencies for 'testFixtures' sources. (n)
testFixturesWearApp - Link to a wear app to embed for object 'testFixtures'. (n)
testImplementation - Implementation only dependencies for 'test' sources. (n)
testReleaseAnnotationProcessor - Classpath for the annotation processor for 'testRelease'. (n)
testReleaseCompileOnly - Compile only dependencies for 'testRelease' sources. (n)
testReleaseImplementation - Implementation only dependencies for 'testRelease' sources. (n)
testReleaseRuntimeOnly - Runtime only dependencies for 'testRelease' sources. (n)
testReleaseWearApp - Link to a wear app to embed for object 'testRelease'. (n)
testRuntimeOnly - Runtime only dependencies for 'test' sources. (n)
testWearApp - Link to a wear app to embed for object 'test'. (n)
wearApp - Link to a wear app to embed for object 'main'. (n)
(c) - A dependency constraint, not a dependency. The dependency affected by the constraint occurs elsewhere in the tree.
(*) - Indicates repeated occurrences of a transitive dependency subtree. Gradle expands transitive dependency subtrees only once per project; repeat occurrences only display the root of the subtree, followed by this annotation.
(n) - A dependency or dependency configuration that cannot be resolved.
```

**Comment by loosebazooka** [comment]
Honestly I think releaseRuntimeClasspath is the one you want. Maybe @mathjeff or @AlexanderYukhanov might have a better idea.

Also it might be helpful to understand who is consuming your sbom?

**Comment by derpeter** [comment]
Using releaseRuntimeClasspath results in

```
A problem occurred configuring root project 'OpenMobileNetworkToolkit'.
> Could not create task ':spdxSbomForRelease'.
   > Configuration with name 'releaseRuntimeClasspath' not found.
```

sorry im very lost here what greadle does and what comes from configurations.

Consumer from the spdx will first Excel sheet (i know i know) but that what the person in charge needs.
On the long run im not sure, maybe depencytrack.

**Comment by loosebazooka** [comment]
Are you configuring it on the app subproject or the root project? I think for Android spdx should be applied on `./app/build gradle.kts`

**Comment by derpeter** [comment]
I tried both, i will do it again tomorrow and post both outputs. In my case ist gradle.gradle so no kotline grade. If that makes a difference.

**Comment by loosebazooka** [comment]
So this is pretty hard for me to debug. If you could share a sample project on github, that could speed this up significantly

**Comment by derpeter** [comment]
I followed your advice and tried again with moving everthing into the app gradle file. Running it with
```
 ./gradlew app:spdxSbomForRelease
```
results in da duplicated key error

i created a minimal android projekt which should match my other project in terms of language and general settings. 

https://github.com/derpeter/spdx-android-test

it runs into the same error just for a different resrouce.

**Comment by derpeter** [comment]
Looking closer to actuall error message
```
Duplicate key androidx.fragment:fragment:1.3.6 (attempted merging values 

/home/user/.gradle/caches/modules-2/files-2.1/androidx.fragment/fragment/1.3.6/6391fc504e7c0f24051ac8b9da3f0ba4d1241dff/fragment-1.3.6.aar and 

/home/user/.gradle/caches/modules-2/files-2.1/androidx.fragment/fragment/1.3.6/6391fc504e7c0f24051ac8b9da3f0ba4d1241dff/fragment-1.3.6.aar

)
```
It looks like it trys to merge identical files

**Comment by derpeter** [comment]
Hey, can i provide any further assistance here? I played with different fresh android projects and could not get it to work.

**Comment by loosebazooka** [comment]
Oh sorry, I was a bit busy, I'll take a look again.

**Comment by loosebazooka** [comment]
So it looks like android configurations surface the same dependency via two different objects that have the same artifact identifier and same file, so: https://github.com/spdx/spdx-gradle-plugin/blob/main/src/main/java/org/spdx/sbom/gradle/utils/SpdxDocumentBuilder.java#L161-L163 just fails. I think there's a simple fix for it. Lemme try it out

**Comment by loosebazooka** [comment]
https://github.com/spdx/spdx-gradle-plugin/pull/66 should fix the problem. Once that is merged, you can try to the local build against your projects.

**Comment by loosebazooka** [comment]
@derpeter it's merged, please try with a local build

**Comment by derpeter** [comment]
I'm having trouble to make gradle look into the local repo. I tried to add it to settings.gradle in the dependencyResolutionManagement block, as well as in the projekt and app gradle file. For some reasons gradle is still not looking into the local repostitory. Where did you add the repo in the example project?

**Comment by loosebazooka** [comment]
So part of the instructions are adding searching local repo to the settings.xml, that should do it?
```
pluginManagement {
 repositories {
     mavenLocal()
     gradlePluginPortal()
 }
}
```

**Comment by derpeter** [comment]
Ah thanks, i was confused by dependencyResolutionManagement. After adding pluginManagement block  gradle is happy and seems to use local version.

runing:  
```./gradlew :spdxSbomForRelease```
results in an spdx file again only containing the project it self.
running:
```./gradlew app:spdxSbomForRelease```
on app level also results in BUILD SUCCESFULL message but now spdx file at all.

**Comment by loosebazooka** [comment]
Did you check `./app/build/spdx/`? That's where it is expected to be

**Comment by derpeter** [comment]
wohoo thanks, i didn't check there and there is a long json file.
Thanks you for all the support and patience with me.

**Comment by loosebazooka** [comment]
np, you helped us catch some bugs! I'll close this for now.

**Comment by loosebazooka** [comment]
I thought we did release something with your fix? We have 2 open prs that are going in. I think we'll do a release right after those

**Comment by PeterHasse** [comment]
At least the last tag is older then the fix. https://plugins.gradle.org/plugin/org.spdx.sbom also points to 0.3 .
Looking forward to the release to glue it all togther in the CI pipline.
Thanks again for all the help

**Comment by loosebazooka** [comment]
cool, I think we're pretty close. I'll do a release today/tomorrow.

**Comment by loosebazooka** [comment]
should be good to go

**Comment by PeterHasse** [comment]
Works thanks. I realized i missed one warning in the past
```
spdx sboms require a version but project: app has no specified version
```
I guess this is again an android thing, the version is part of the gradle file and can be found in:

```
android {
    defaultConfig {
        applicationId "XYZ"
        minSdk 29
        targetSdk 34
        versionCode 1
        versionName "1.0"
```

maybe this could be included in future version. maybe also the applicationId it worth including

**Comment by loosebazooka** [comment]
Yeah we might need special handling for android

**Comment by loosebazooka** [comment]
Can you open a new issue?
